From 7f9663fc1c7c198486e5ea1f4e72c29575c3f89f Mon Sep 17 00:00:00 2001
From: Andi Kleen <andi@firstfloor.org>
Date: Tue, 7 Jul 2020 09:59:36 -0700
Subject: [PATCH 29/38] kbuild, kallsyms: Add sanity check for offsets

The single link kallsyms relies on all the symbols in
the first pass and second pass kallsyms matching.

Add a sanity check during build that this is the case.

Signed-off-by: Andi Kleen <ak@linux.intel.com>
---
 scripts/link-vmlinux.sh | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/scripts/link-vmlinux.sh b/scripts/link-vmlinux.sh
index bce23081b0be..32ec1a82a58a 100755
--- a/scripts/link-vmlinux.sh
+++ b/scripts/link-vmlinux.sh
@@ -407,6 +407,12 @@ if [ -n "${CONFIG_KALLSYMS}" -a -n "${CONFIG_KALLSYMS_SINGLE}" ] ; then
 	kallsyms_s vmlinux .tmp_kallsyms2.S `cat .kallsyms_pad`
 	kallsyms_o .tmp_kallsyms2.S .tmp_kallsyms2.o -DNO_REL
 
+	# sanity check the offsets
+	${NM} .tmp_kallsyms1.o >.tmp_kallsyms1.nm
+	${NM} .tmp_kallsyms2.o >.tmp_kallsyms2.nm
+	cmp .tmp_kallsyms1.nm .tmp_kallsyms2.nm
+	rm .tmp_kallsyms[12].nm
+
 	info OBJCOPY .tmp_kallsyms2.bin
 	${OBJCOPY} -O binary .tmp_kallsyms2.o .tmp_kallsyms2.bin
 
-- 
2.28.0

